src/gp_doctest.cpp

```diff
@@ -766,3 +767,81 @@ TEST_CASE("GPInstance: Only add child node tests") {
   CHECK_EQ(dag.GetEdgeRange(dag.GetDAGNode(11)->GetBitset(), false).first, 3);
   CHECK_EQ(dag.GetEdgeRange(dag.GetDAGNode(11)->GetBitset(), false).second, 5);
 }
+
+// See before adding NNI diagram at:
+// https://user-images.githubusercontent.com/31897211/133525012-dd46feed-e11d-44c8-9297-744a3e6314c2.gif
+// See after adding NNI diagram at:
+// https://user-images.githubusercontent.com/31897211/134055844-2c4df31f-7896-4b7f-85c9-782028e29064.gif
+TEST_CASE("GPInstance: SubsplitDAG NNI (Nearest Neighbor Interchange)") {
+  // Simple sDAG that contains a shared edge, internal leafward fork, and an internal
+  // rootward fork.
+  const std::string fasta_path = "data/six_taxon.fasta";
+  auto inst = GPInstanceOfFiles(fasta_path, "data/six_taxon_rooted_simple.nwk");
+  SubsplitDAG& dag = inst.GetDAG();
+  // Build NNI Set from current DAG state.
+  auto set_of_nnis = dag.BuildSetOfNNIs();
+  auto correct_set_of_nnis = SetOfNNIs();
+
+  // Functions for quick insertion/removal to NNI Set.
+  auto InsertNNI = [&](Bitset parent, Bitset child) {
+    auto nni = NNIOperation(parent, child);
+    correct_set_of_nnis.Insert(nni);
+  };
+  auto RemoveNNI = [&](Bitset parent, Bitset child) {
+    auto nni = NNIOperation(parent, child);
+    correct_set_of_nnis.Erase(nni);
+  };
+
+  // Add NNI's for edge 4 to NNI Set.
+  InsertNNI(Bitset::Subsplit("010000", "101111"),
```

Comment:
I think it would be most fun to sit down and work through these together. Let's do that.